rust/src/builders/input_builder.rs

```diff
@@ -0,0 +1,161 @@
+use crate::*;
+use crate::builders::witness_builder::{InputAggregateWitnessData, PartialPlutusWitness};
+
+use super::witness_builder::{RequiredWitnessSet, NativeScriptWitnessInfo, PlutusScriptWitnessInfo};
+
+pub fn input_required_wits(utxo_info: &TransactionOutput, required_witnesses: &mut RequiredWitnessSet) -> () {
+    if let Some(cred) = &utxo_info.address().payment_cred() {
+        if let Some(keyhash) = &cred.to_keyhash() {
+            required_witnesses.add_vkey_key_hash(&keyhash);
+        }
+        if let Some(script_hash) = &cred.to_scripthash() {
+            required_witnesses.add_script_hash(&script_hash);
+            if let Some(data_hash) = utxo_info.data_hash() {
+                required_witnesses.add_plutus_datum_hash(&data_hash);
+                // note: redeemer is required as well
+                // but we can't know the index, so we rely on the tx builder to satisfy this requirement
+            }
+        }
+    };
+}
+
+#[wasm_bindgen]
+#[derive(Clone)]
+pub struct InputBuilderResult {
+    input: TransactionInput,
+    utxo_info: TransactionOutput,
+    aggregate_witness: Option<InputAggregateWitnessData>,
+    required_wits: RequiredWitnessSet,
+}
+
+#[wasm_bindgen]
+#[derive(Clone)]
+pub struct SingleInputBuilder {
+    input: TransactionInput,
+    utxo_info: TransactionOutput,
+}
+
+#[wasm_bindgen]
+impl SingleInputBuilder {
+    pub fn new(input: &TransactionInput, utxo_info: &TransactionOutput,) -> Self {
+        Self {
+            input: input.clone(),
+            utxo_info: utxo_info.clone(),
+        }
+    }
+
+    pub fn skip_witness(&self) -> Result<InputBuilderResult, JsError> {
+        let mut required_wits = RequiredWitnessSet::default();
+        input_required_wits(&self.utxo_info, &mut required_wits);
+
+        Ok(InputBuilderResult {
+            input: self.input.clone(),
+            utxo_info: self.utxo_info.clone(),
+            aggregate_witness: None,
+            required_wits: required_wits.clone(),
+        })
+    }
+
+    pub fn vkey(&self, vkey: &Vkey) -> Result<InputBuilderResult, JsError> {
+        let mut required_wits = RequiredWitnessSet::default();
+        input_required_wits(&self.utxo_info,&mut required_wits);
+        let mut required_wits_left = required_wits.clone();
+
+        let keyhash = &vkey.public_key().hash();
```

Comment:
it's accessed everywhere as &keyhash so we should either access it as `keyhash` directly (as it's a ref - rust magically saves temps that you take a ref of like that from going out of scope) or make it not a ref. I think this is a clippy warning too? Or one of the rust static checkers.